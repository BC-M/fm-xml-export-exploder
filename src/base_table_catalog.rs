use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;
use std::collections::HashMap;
use std::io::{BufRead, Read};

use crate::utils::attributes::get_attributes;

pub fn parse_base_table_catalog<R: Read + BufRead>(
    reader: &mut Reader<R>,
    _: &BytesStart,
) -> HashMap<String, String> {
    let mut depth = 1;
    let mut table_name_id_map: HashMap<String, String> = HashMap::new();

    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error {}", e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;

                if depth == 2 && e.name().as_ref() == b"BaseTable" {
                    let mut id = String::new();
                    let mut name = String::new();

                    for attr in get_attributes(&e).unwrap() {
                        match attr.0.as_str() {
                            "id" => id = attr.1.to_string(),
                            "name" => name = attr.1.to_string(),
                            _ => {}
                        }
                    }

                    table_name_id_map.insert(name, id);
                }
            }
            Ok(Event::End(_)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }
            }
            _ => {}
        }

        buf.clear()
    }

    table_name_id_map
}
