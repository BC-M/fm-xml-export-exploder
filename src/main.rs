use quick_xml::events::Event;
use quick_xml::reader::Reader;

use crate::base_table_catalog::parse_base_table_catalog;
use crate::custom_function_catalog::xml_explode_custom_function_catalog;
use crate::external_data_source_catalog::xml_extract_external_data_sources;
use crate::layout_catalog::xml_explode_layout_catalog;
use crate::relationship_catalog::xml_explode_relationship_catalog;
use crate::script_catalog::parse_script_directories;
use crate::script_steps_catalog::xml_explode_script_catalog;
use crate::table_catalog::xml_explode_table_catalog;
use crate::theme_catalog::xml_explode_theme_catalog;
use crate::utils::attributes::get_attribute;
use crate::value_list_catalog::xml_explode_value_list_catalog;
use encoding_rs_io::DecodeReaderBytes;
use rayon::prelude::*;
use std::collections::HashMap;
use std::{
    env,
    error::Error,
    fs,
    fs::File,
    io::BufReader,
    path::{Path, PathBuf},
    time::Instant,
};

mod base_table_catalog;
mod calculations;
mod custom_function_catalog;
mod external_data_source_catalog;
mod layout_catalog;
mod relationship_catalog;
mod script_catalog;
mod script_steps;
mod script_steps_catalog;
mod table_catalog;
mod theme_catalog;
mod utils;
mod value_list_catalog;

fn main() -> Result<(), Box<dyn Error>> {
    let start = Instant::now();

    let in_dir = env::args().nth(1).ok_or("Input directory not provided")?;
    let out_dir = env::args().nth(2).ok_or("Output directory not provided")?;

    let in_dir = Path::new(&in_dir);
    if !is_valid_directory(in_dir) {
        println!("'{}' is not a valid directory.", in_dir.display());
        return Ok(());
    }

    let out_dir = Path::new(&out_dir);
    if !is_valid_directory(out_dir) {
        println!("'{}' is not a valid directory.", out_dir.display());
        return Ok(());
    }

    // Read directory contents
    let paths = fs::read_dir(in_dir)?
        .filter_map(|entry| entry.ok().map(|e| e.path())) // Filter out directories and unwrap results
        .filter(|path| path.is_file() && path.extension().unwrap_or_default() == "xml") // Filter XML files
        .collect::<Vec<_>>(); // Collect paths into a vector

    println!("Start processing {} files...", paths.len());

    // Process XML files in parallel
    paths.par_iter().for_each(|path| explode_xml(path, out_dir));

    let duration = start.elapsed();
    if duration.as_secs() > 9 {
        println!("Completed in {:?} seconds.", duration.as_secs());
    } else {
        println!("Completed in {:?} ms.", duration.as_millis());
    }

    Ok(())
}

fn is_valid_directory(dir_path: &Path) -> bool {
    fs::metadata(dir_path)
        .map(|metadata| metadata.is_dir())
        .unwrap_or(false)
}

fn explode_xml(fm_export_file_path: &PathBuf, out_dir_path: &Path) {
    let start = Instant::now();
    let fm_export_file_name = fm_export_file_path.file_name().unwrap().to_str().unwrap();

    // Open XML file
    let file = match File::open(fm_export_file_path) {
        Ok(file) => file,
        Err(err) => {
            eprintln!(
                "Error opening file {}: {}",
                fm_export_file_path.display(),
                err
            );
            return;
        }
    };

    // Initialize variables
    let mut fm_file_name = String::new();
    let mut depth = 0;
    let mut script_id_path_map: HashMap<String, Vec<String>> = HashMap::new();
    let mut table_name_id_map: HashMap<String, String> = HashMap::new();

    // Iterate over XML events
    let decode_reader = BufReader::new(DecodeReaderBytes::new(file));
    let mut reader = Reader::from_reader(decode_reader);
    let mut buf: Vec<u8> = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error in {}: {}", fm_export_file_path.display(), e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                match depth {
                    0 => match e.name().as_ref() {
                        b"FMDynamicTemplate" | b"FMSaveAsXML" => {
                            fm_file_name = get_attribute(&e, "File")
                                .unwrap()
                                .strip_suffix(".fmp12")
                                .unwrap()
                                .to_string()
                        }
                        _ => {
                            eprintln!("Unsupported XML-format in file {}", fm_export_file_name);
                            return;
                        }
                    },
                    2 => {
                        if e.name().as_ref() == b"ModifyAction" {
                            break;
                        }
                    }
                    3 => match e.name().as_ref() {
                        b"BaseTableCatalog" => {
                            table_name_id_map = parse_base_table_catalog(&mut reader, &e);
                            continue;
                        }
                        b"LayoutCatalog" => {
                            xml_explode_layout_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"FieldsForTables" => {
                            xml_explode_table_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                                &table_name_id_map,
                            );
                            continue;
                        }
                        b"CalcsForCustomFunctions" => {
                            xml_explode_custom_function_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"StepsForScripts" => {
                            xml_explode_script_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                                &script_id_path_map,
                            );
                            continue;
                        }
                        b"ScriptCatalog" => {
                            script_id_path_map = parse_script_directories(&mut reader, &e);
                            continue;
                        }
                        b"ExternalDataSourceCatalog" => {
                            xml_extract_external_data_sources(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"ValueListCatalog" => {
                            xml_explode_value_list_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"RelationshipCatalog" => {
                            xml_explode_relationship_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"ThemeCatalog" => {
                            xml_explode_theme_catalog(&mut reader, &e, out_dir_path, &fm_file_name);
                            continue;
                        }
                        _ => {}
                    },
                    _ => {}
                }
                depth += 1;
            }
            Ok(Event::End(_)) => {
                depth -= 1;
            }
            _ => {}
        }
        buf.clear()
    }

    println!(
        "{} finished in {} ms.",
        fm_export_file_name,
        start.elapsed().as_millis()
    );
}

fn join_scope_id_and_name(scope_id: &str, scope_name: &str) -> String {
    format!("{} - ID {}", scope_name, scope_id)
}

fn escape_filename(filename: &str) -> String {
    // Replace special characters with underscores
    let escaped_filename = filename
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '{' => '_',
            _ => c,
        })
        .collect::<String>();

    escaped_filename.trim().to_string()
}

fn should_skip_line(line: &str) -> bool {
    // de-noise TODO: make optional
    if line.contains("<TagList></TagList>") {
        return true;
    };
    if line.contains("<OwnerID></OwnerID>") {
        return true;
    };
    if line.contains("<Options>0</Options>") {
        return true;
    }; // Default
    if line.contains("<Options>1048576</Options>") {
        return true;
    }; // Breakpoint
    if line.contains("<UUID>") && line.contains("</UUID>") {
        return true;
    };

    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use fs::remove_dir_all;
    use insta::assert_snapshot;
    use similar_asserts::assert_eq;
    use std::io::Read;
    use walkdir::WalkDir;

    #[test]
    fn test_escape_filename() {
        assert_eq!(escape_filename("filename.xml"), "filename.xml");
        assert_eq!(escape_filename("file/name.xml"), "file_name.xml");
        assert_eq!(escape_filename("file|name.xml"), "file_name.xml");
    }

    #[test]
    fn snapshot_test() {
        let snapshot_dir = Path::new("./tests/snapshots");
        let input_dir = Path::new("./tests/xml");
        let output_dir = Path::new("./out");
        let _ = remove_dir_all(output_dir);

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path(snapshot_dir);
        settings.set_prepend_module_to_snapshot(false);

        let paths = fs::read_dir(input_dir)
            .unwrap()
            .filter_map(|entry| entry.ok().map(|e| e.path()))
            .filter(|path| path.is_file())
            .filter(|e| e.file_name().unwrap() != ".DS_Store")
            .collect::<Vec<_>>();

        for path in paths {
            explode_xml(&path, output_dir);
        }

        let output_files: Vec<PathBuf> = WalkDir::new(output_dir)
            .into_iter()
            .filter_map(|entry| entry.ok())
            .filter(|e| e.path().is_file())
            .filter(|e| e.file_name().to_str().unwrap() != ".DS_Store")
            .map(|e| e.path().to_path_buf())
            .collect();

        let mut output_file_paths = output_files
            .iter()
            .map(|file| file.strip_prefix(output_dir).unwrap())
            .map(|file| file.to_string_lossy())
            .collect::<Vec<_>>();
        output_file_paths.sort();

        let snapshot_files: Vec<PathBuf> = WalkDir::new(snapshot_dir)
            .into_iter()
            .filter_map(|entry| entry.ok())
            .filter(|e| e.path().is_file())
            .filter(|e| e.file_name().to_str().unwrap().ends_with(".snap"))
            .map(|e| e.path().to_path_buf())
            .collect();

        let mut snapshot_file_paths = snapshot_files
            .iter()
            .map(|file| file.strip_prefix(snapshot_dir).unwrap())
            .map(|file| {
                file.to_string_lossy()
                    .strip_suffix(".snap")
                    .unwrap()
                    .to_string()
            })
            .collect::<Vec<_>>();
        snapshot_file_paths.sort();

        assert_eq!(output_file_paths.join("\n"), snapshot_file_paths.join("\n"));

        for output_file in &output_files {
            let output_content = String::from_utf8(read_file(output_file)).unwrap();
            let output_file_slug = output_file.strip_prefix(output_dir).unwrap();

            settings.set_snapshot_path(
                Path::new("../tests/snapshots")
                    .join(output_file_slug)
                    .parent()
                    .unwrap(),
            );

            settings.bind(|| {
                assert_snapshot!(output_file.file_name().unwrap().to_str(), output_content);
            });
        }
    }

    fn read_file(file_path: &PathBuf) -> Vec<u8> {
        let mut file = fs::File::open(file_path).expect("Failed to open file");
        let mut content = Vec::new();
        file.read_to_end(&mut content).expect("Failed to read file");
        content
    }
}
