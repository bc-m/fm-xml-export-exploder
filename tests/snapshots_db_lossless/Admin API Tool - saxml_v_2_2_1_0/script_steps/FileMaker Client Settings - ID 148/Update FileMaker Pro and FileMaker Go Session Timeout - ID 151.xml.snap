---
source: src/main.rs
expression: output_content
---
<Script>
	<ScriptReference id="151" name="Update FileMaker Pro and FileMaker Go Session Timeout" UUID="7E913F98-12B6-4A58-9B9A-CF01FEF3A37F"></ScriptReference>
	<ObjectList membercount="33">
		<Step index="0" id="89" name="# (Kommentar)" enable="True">
			<UUID>71A371A3-412A-414A-B3E8-D4FB532DA2A7</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" PURPOSE: Update FileMaker Pro and FileMaker Go Session Timeout"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="1" id="89" name="# (Kommentar)" enable="True">
			<UUID>8DB6FBB6-61D1-446C-860A-E684E5DEAB5C</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" HISTORY:&#13;   - Created:   2018-05-09 by Mike Duncan (mduncan@soliantconsulting.com)&#13;   - Modified:  2019-04-23 by Mike Duncan, update to v2 admin api"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="2" id="89" name="# (Kommentar)" enable="True">
			<UUID>E81732B1-2BD7-4BE3-A2A6-066B2BD5D140</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="3" id="85" name="AnwenderAbbruchZulassen setzen" enable="True">
			<UUID>6B421C36-71B4-4DBD-B14B-F9E0C13E32A0</UUID>
			<OwnerID></OwnerID>
			<Options>196608</Options>
			<ParameterValues membercount="1">
				<Parameter type="Boolean">
					<Boolean id="131072" value="True"></Boolean>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="4" id="86" name="Fehleraufzeichnung setzen" enable="True">
			<UUID>9F4AD967-43AA-4173-AA6F-470B610F78A2</UUID>
			<OwnerID></OwnerID>
			<Options>196608</Options>
			<ParameterValues membercount="1">
				<Parameter type="Boolean">
					<Boolean id="131072" value="True"></Boolean>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="5" id="141" name="Variable setzen" enable="True">
			<UUID>D4503A55-2DCF-49AC-BA04-BA7CF74B09F8</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[0]]></Text>
								<ChunkList hash="9BF55A5D51230A8EA20D4BAA0F468EEC">
									<Chunk type="NoRef">0</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$error"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="6" id="89" name="# (Kommentar)" enable="True">
			<UUID>6E4F4AA8-7C11-496E-BB29-C23B56AB132D</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="7" id="89" name="# (Kommentar)" enable="True">
			<UUID>36987C31-FCA8-4E3D-B7BB-39EF95FBECF0</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="8" id="89" name="# (Kommentar)" enable="True">
			<UUID>FF4F0C74-2D1E-4F23-83D9-DD779D28FAAD</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" SET-UP"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="9" id="89" name="# (Kommentar)" enable="True">
			<UUID>30BA25B6-6296-4C0C-9021-A5909F4F8F23</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="10" id="1" name="Script ausfÃ¼hren" enable="True">
			<UUID>12DFDFB3-F06A-4B8E-A7B0-3B9A7E223803</UUID>
			<OwnerID></OwnerID>
			<Options>64</Options>
			<ParameterValues membercount="2">
				<Parameter type="List">
					<List name="Aus Liste" value="1">
						<ScriptReference id="35" name="Set Config" UUID="6E806D84-9200-44AE-B032-89A6016326FF"></ScriptReference>
					</List>
				</Parameter>
				<Parameter type="Parameter">
					<Parameter></Parameter>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="11" id="141" name="Variable setzen" enable="True">
			<UUID>A3B99543-3F01-42EB-8767-8FEE94E58C56</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA["/fmclients/sessiontimeout/progo"]]></Text>
								<ChunkList hash="4C69754A11CC30F4F2A4EE1E8A832D9F">
									<Chunk type="NoRef">&quot;/fmclients/sessiontimeout/progo&quot;</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$endpoint"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="12" id="141" name="Variable setzen" enable="True">
			<UUID>4C9BAC3D-A3D7-47C6-9EB9-F27C4FF98745</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[JSONSetElement ( "" ;[ "enableProGoTimeout"; FMS__fmserver::enableProGoTimeout ; JSONBoolean ] ;[ "sessionTimeout"; FMS__fmserver::sessionTimeout_progo ; JSONNumber ])]]></Text>
								<ChunkList hash="A7204E2570AC907973528E9256C92518">
									<Chunk type="FunctionRef">JSONSetElement</Chunk>
									<Chunk type="NoRef"> ( &quot;&quot;&#13; ;[ &quot;enableProGoTimeout&quot;; </Chunk>
									<Chunk type="FieldRef">
										<FieldReference id="149" name="enableProGoTimeout" repetition="1" UUID="CA5B2875-FA57-4761-A66A-B783BE8F0E7C">
											<TableOccurrenceReference id="1065089" name="FMS__fmserver" UUID="597569C5-24CD-486D-84D6-D79B9D3BDD5B"></TableOccurrenceReference>
										</FieldReference>
									</Chunk>
									<Chunk type="NoRef"> ; </Chunk>
									<Chunk type="FunctionRef">JSONBoolean</Chunk>
									<Chunk type="NoRef"> ]&#13; ;[ &quot;sessionTimeout&quot;; </Chunk>
									<Chunk type="FieldRef">
										<FieldReference id="150" name="sessionTimeout_progo" repetition="1" UUID="C416306F-4DD8-47AC-BCDD-1F7AC8443795">
											<TableOccurrenceReference id="1065089" name="FMS__fmserver" UUID="597569C5-24CD-486D-84D6-D79B9D3BDD5B"></TableOccurrenceReference>
										</FieldReference>
									</Chunk>
									<Chunk type="NoRef"> ; </Chunk>
									<Chunk type="FunctionRef">JSONNumber</Chunk>
									<Chunk type="NoRef"> ]&#13;)</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$json"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="13" id="141" name="Variable setzen" enable="True">
			<UUID>41AE8915-C655-4630-9DF2-11D274A8646D</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[" --request PATCH" &" --header \"Authorization: Bearer " & $$my.token & "\"" &" --header \"Content-Type: application/json\"" &" --data @$json" &" --dump-header $$my.headers"]]></Text>
								<ChunkList hash="6647F483744AACEB16B9320A2A818240">
									<Chunk type="NoRef">&quot; --request PATCH&quot; &amp;&#13;&quot; --header \&quot;Authorization: Bearer &quot; &amp; </Chunk>
									<Chunk type="VariableReference">$$my.token</Chunk>
									<Chunk type="NoRef"> &amp; &quot;\&quot;&quot; &amp;&#13;&quot; --header \&quot;Content-Type: application/json\&quot;&quot; &amp;&#13;&quot; --data @$json&quot; &amp;&#13;&quot; --dump-header $$my.headers&quot;</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$curloptions"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="14" id="89" name="# (Kommentar)" enable="True">
			<UUID>0A5A7E7D-2B86-4301-81A8-4CE13F9B6E46</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="set api update variable"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="15" id="141" name="Variable setzen" enable="True">
			<UUID>80924E62-5065-4C4D-B270-661E7B58150E</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA["1"]]></Text>
								<ChunkList hash="E42599A078EF5CBB8E2C8EC3C24F3514">
									<Chunk type="NoRef">&quot;1&quot;</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$$API.UPDATE"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="16" id="89" name="# (Kommentar)" enable="True">
			<UUID>4BF24233-3F51-4951-87F2-4411DDD65D10</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="17" id="89" name="# (Kommentar)" enable="True">
			<UUID>2F73CECB-C825-4841-844C-C6C785241CCE</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="18" id="89" name="# (Kommentar)" enable="True">
			<UUID>4CE85DC1-0503-47E0-865E-EA1148AC814F</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" PROCESS"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="19" id="89" name="# (Kommentar)" enable="True">
			<UUID>B43FCBD4-CAFC-4819-860D-F66C42B24193</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="20" id="160" name="Aus URL einfÃ¼gen" enable="True">
			<UUID>2954BBD4-EAFE-46E8-92BA-A8106E8C60CC</UUID>
			<OwnerID></OwnerID>
			<Options>20613</Options>
			<ParameterValues membercount="6">
				<Parameter type="Boolean">
					<Boolean type="SSL-Zertifikate verifizieren" id="268435456" value="False"></Boolean>
				</Parameter>
				<Parameter type="Boolean">
					<Boolean type="Auswahl" id="4096" value="True"></Boolean>
				</Parameter>
				<Parameter type="Boolean">
					<Boolean type="Mit Dialog" id="128" value="False"></Boolean>
				</Parameter>
				<Parameter type="Target">
					<Variable value="$$my.result">
						<repetition value="1"></repetition>
					</Variable>
				</Parameter>
				<Parameter type="URL">
					<URL autoEncode="True">
						<Calculation datatype="1" position="0">
							<Calculation>
								<Text><![CDATA[$$server & $endpoint]]></Text>
								<ChunkList hash="9EA92AEF077FA44AE68308694556F8CB">
									<Chunk type="VariableReference">$$server</Chunk>
									<Chunk type="NoRef"> &amp; </Chunk>
									<Chunk type="VariableReference">$endpoint</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</URL>
				</Parameter>
				<Parameter type="Calculation">
					<Calculation datatype="1" position="1">
						<Calculation>
							<Text><![CDATA[$curloptions]]></Text>
							<ChunkList hash="205B4DA1BAD160DEDA31FE70167BBA78">
								<Chunk type="VariableReference">$curloptions</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="21" id="1" name="Script ausfÃ¼hren" enable="True">
			<UUID>DB74DC4C-A043-4C2E-B53B-5E93CF53F224</UUID>
			<OwnerID></OwnerID>
			<Options>64</Options>
			<ParameterValues membercount="2">
				<Parameter type="List">
					<List name="Aus Liste" value="1">
						<ScriptReference id="31" name="Handle response" UUID="44FA18ED-6DEA-4F32-BDDB-09F438F6D005"></ScriptReference>
					</List>
				</Parameter>
				<Parameter type="Parameter">
					<Parameter></Parameter>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="22" id="89" name="# (Kommentar)" enable="True">
			<UUID>413A210A-BD54-4E50-A709-05BBB9EE102B</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="23" id="89" name="# (Kommentar)" enable="True">
			<UUID>8A615C41-3E85-43F5-9024-E5597E43139F</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="24" id="89" name="# (Kommentar)" enable="True">
			<UUID>94EF5C36-5431-4894-83E0-68DA1DD9FD28</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" CLEAN UP"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="25" id="89" name="# (Kommentar)" enable="True">
			<UUID>22D2F341-8104-4C1E-9741-16F2C0D4C25E</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value="-------------------------------------------------------------"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="26" id="68" name="Wenn" enable="True">
			<UUID>A0E1508E-EF12-4073-83A5-110F5064D7A9</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="7" position="0">
						<Calculation>
							<Text><![CDATA[$error = 0]]></Text>
							<ChunkList hash="CE3C173E2CD2554CBB236CEE7C56CBB6">
								<Chunk type="VariableReference">$error</Chunk>
								<Chunk type="NoRef"> = 0</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="27" id="89" name="# (Kommentar)" enable="True">
			<UUID>58069074-9519-430D-A600-69796D522B90</UUID>
			<OwnerID></OwnerID>
			<Options>4</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment value=" set to exit true if you do not need to pass an error"></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="28" id="141" name="Variable setzen" enable="True">
			<UUID>CA84533C-D13D-4D8C-B95B-4E88DE6EE9F0</UUID>
			<OwnerID></OwnerID>
			<Options>16388</Options>
			<ParameterValues membercount="1">
				<Parameter type="Variable">
					<value>
						<Calculation datatype="1" position="1">
							<Calculation>
								<Text><![CDATA[1 // exit true]]></Text>
								<ChunkList hash="DC134DF247804ECF23193E2024A59A1B">
									<Chunk type="NoRef">1 // exit true</Chunk>
								</ChunkList>
							</Calculation>
						</Calculation>
					</value>
					<Name value="$error"></Name>
					<repetition></repetition>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="29" id="70" name="Ende (wenn)" enable="True">
			<UUID>6CBDF4C2-6F0D-49FC-9E18-1E60196C204B</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
		</Step>
		<Step index="30" id="89" name="# (Kommentar)" enable="True">
			<UUID>45607ADB-478F-4A35-A395-3B957506EB9D</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="31" id="103" name="Aktuelles Script verlassen" enable="True">
			<UUID>8C61FEF8-E21D-4C9E-BEA2-3399D7FF9AAD</UUID>
			<OwnerID></OwnerID>
			<Options>16384</Options>
			<ParameterValues membercount="1">
				<Parameter type="Calculation">
					<Calculation datatype="1" position="0">
						<Calculation>
							<Text><![CDATA[$error]]></Text>
							<ChunkList hash="926F59D0E89012933FB1C2137ABF78F1">
								<Chunk type="VariableReference">$error</Chunk>
							</ChunkList>
						</Calculation>
					</Calculation>
				</Parameter>
			</ParameterValues>
		</Step>
		<Step index="32" id="89" name="# (Kommentar)" enable="True">
			<UUID>C64F52C0-36D3-43D8-AB07-538A9C2A00C4</UUID>
			<OwnerID></OwnerID>
			<Options>0</Options>
			<ParameterValues membercount="1">
				<Parameter type="Comment">
					<Comment></Comment>
				</Parameter>
			</ParameterValues>
		</Step>
	</ObjectList>
</Script>
